### ISSUE TITLE
Union types and type intersections don't work together

### ISSUE THREAD

**Issue by kalmi** [vega/ts-json-schema-generator#36]
Hello,

I am investigating whether we could use this library to generate json schemas from our existing TypeScript interfaces. I have run into an issue with union types and type intersections. Is this a known limitation? In the provided example the "c" field is not present at all in the output.

input (test.ts):
```
interface A {
  x: boolean;
  c: 'b';
}
interface B {
  x: boolean;
  c: 'c';
}
export type D = (A|B) & {x: true};
```
`ts-json-schema-generator --strictTuples --path 'test.ts' --expose 'export' --type D`

output:
```
{
  "$ref": "#/definitions/D",
  "$schema": "http://json-schema.org/draft-06/schema#",
  "definitions": {
    "D": {
      "additionalProperties": false,
      "properties": {
        "x": {
          "enum": [
            true
          ],
          "type": "boolean"
        }
      },
      "required": [
        "x"
      ],
      "type": "object"
    }
  }
}
```

Thanks,
Kálmán

**Comment by kayahr** [comment]
Ran into the same problem...

```typescript
type Base = {
    a: string;
};

type Variant1 = {
    b: string;
};

type Variant2 = {
    b: number;
};

export type MyObject = Base & (Variant1 | Variant2);
```

It works when restructuring the type a little bit, so this may be used as a workaround:

```typescript
export type MyObject = (Base & Variant1) | (Base & Variant2);
```

**Comment by domoritz** [comment]
We should be doing this automatically but it's not trivial to implement with the current architecture.

**Comment by kayahr** [comment]
There seems to be another issue with the original example. Even when translating the intersection type manually to `A & {x: true} | B & {x: true}` it does not yield the correct schema because the restriction of `x` to `true` is ignored. It works when swapping the order to `{x: true} & A | {x: true} & B`. This has nothing to do with unions, can also be reproduced with a simple intersection. I'll create a new issue for this.

**Comment by kayahr** [comment]
I'm closing this issue because the main part of it is fixed now by #116. The still remaining issue with objects not being properly merged together is tracked by issue #115.
